eventstream.go

```diff
@@ -218,3 +330,86 @@ func (e *EventStream) SubscriberSpecificEvent(f func(e Event), events ...interfa
 	}
 	return &s
 }
+
+// SubscriberAllUntyped bind the eventF function to be called on all events independent on aggregate or event type
+func (e *EventStream) SubscriberAllUntyped(f func(e EventUntyped)) *subscriptionEventUntyped {
+	s := subscriptionEventUntyped{
+		eventF: f,
+	}
+	s.unsubF = func() {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		for i, sub := range e.allEventsUntyped {
+			if &s == sub {
+				e.allEventsUntyped = append(e.allEventsUntyped[:i], e.allEventsUntyped[i+1:]...)
+				break
+			}
+		}
+	}
+	s.subF = func() {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		e.allEventsUntyped = append(e.allEventsUntyped, &s)
+	}
+	return &s
+}
+
+// SubscriberAggregateTypeUntyped bind the eventF function to be called on events on the aggregate type
+func (e *EventStream) SubscriberAggregateTypeUntyped(f func(e EventUntyped), aggregateTypes ...string) *subscriptionEventUntyped {
+	s := subscriptionEventUntyped{
+		eventF: f,
+	}
+	s.unsubF = func() {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		for _, ref := range aggregateTypes {
+			for i, sub := range e.aggregateTypesUntyped[ref] {
+				if &s == sub {
+					e.aggregateTypesUntyped[ref] = append(e.aggregateTypesUntyped[ref][:i], e.aggregateTypesUntyped[ref][i+1:]...)
+					break
+				}
+			}
+		}
+	}
+	s.subF = func() {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		for _, ref := range aggregateTypes {
+			e.aggregateTypesUntyped[ref] = append(e.aggregateTypesUntyped[ref], &s)
+		}
+	}
+	return &s
+}
+
+// SubscriberEventReasonUntyped bind the eventF function to be called on event reason
+func (e *EventStream) SubscriberEventReasonUntyped(f func(e EventUntyped), eventReasons ...string) *subscriptionEventUntyped {
+	s := subscriptionEventUntyped{
+		eventF: f,
+	}
+	s.unsubF = func() {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		for _, ref := range eventReasons {
+			for i, sub := range e.eventReasonUntyped[ref] {
+				if &s == sub {
+					e.eventReasonUntyped[ref] = append(e.eventReasonUntyped[ref][:i], e.eventReasonUntyped[ref][i+1:]...)
+					break
+				}
+			}
+		}
+	}
+	s.subF = func() {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		for _, ref := range eventReasons {
+			e.eventReasonUntyped[ref] = append(e.eventReasonUntyped[ref], &s)
+		}
+	}
+	return &s
+}
```

Comment:
It appears there's a lot of duplicated code in these two adjacent functions. Could it be refactored somehow, or would that require to use generics?

Reply:
Agree but has not find a nice was yet. Generics could be the way but it will force users of the package to also use go 1.18.